### PR TITLE
Update Testerina guides 

### PR DESCRIPTION
**Pull request by Dilhasha**
## Purpose
Update Testerina guides for SL GA release

## Check List

- [ ] **Page Addition**
  - [ ] Add `permalink` to pages
  - [ ] If contains empty folder(s), Add front-matter `redirect_to:`

- [ ] **Page Rename**
  - [ ] Add front-matter `redirect_from`
  - [ ] Add front-matter `redirect_to:` (If applicable)

### CODE REVIEW

File: _data/guidesnavswanlake.yml [hunk 1]
```diff
@@ -39,6 +39,9 @@
     - title: Writing Tests
       url: /learn/testing-ballerina-code/writing-tests/
       active: writing-tests
+    - title: Structuring the tests
```
Comment (praneesha) [inline _data/guidesnavswanlake.yml:42]:
```suggestion
    - title: Structuring the Tests
```
Please use title case for all titles

File: learn/guides/testing-ballerina-code/writing-tests.md [hunk 2]
```diff
@@ -43,15 +29,23 @@ Once the test module is imported, the following annotation can be used to write
 
 ### '@test:Config {}'
 
-The function specified after the annotation is a test function. This annotation supports the following value fields.
+The function specified after the annotation is a test function. This annotation supports fields to configure the test case with the following.
+
+*   Enable/disable the test.
+    ***enable: {true&#124;false}*** - Default: true
+    
+*   Setup and tear down prerequisites required for a test case.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/writing-tests.md:37]:
```suggestion
*   Set up and tear down the prerequisites required for a test case.
```

File: learn/guides/testing-ballerina-code/writing-tests.md [hunk 3]
```diff
@@ -186,7 +174,7 @@ The Ballerina test framework supports the following assertions, which help to ve
     </tr>
 </table>
 
-### Difference between expected and actual values when using 'assertEquals'
+### Troubleshooting assertion failures when using 'assertEquals'
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/writing-tests.md:177]:
```suggestion
### Troubleshooting Assertion Failures when Using 'assertEquals'
```

File: learn/guides/testing-ballerina-code/writing-tests.md [hunk 4]
```diff
@@ -186,7 +174,7 @@ The Ballerina test framework supports the following assertions, which help to ve
...
 
 
 #### Values with different types
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/writing-tests.md:180]:
```suggestion
#### Values with Different Types
```
Please check the title case in all the titles.

File: _data/testingcodeswanlake.yml [hunk 5]
```diff
@@ -67,24 +94,15 @@
         - title: Mocking Functions
           url: /learn/testing-ballerina-code/mocking/#mocking-functions
           active: mocking-functions
-        - title: What’s Next?
-          url: /learn/testing-ballerina-code/mocking/#whats-next
-          active: whats-next
     - title: Executing Tests
       url: '#'
       sublinks:
+        - title: Test Execution behavior
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:100]:
```suggestion
        - title: Test Execution Behavior
```

File: _data/testingcodeswanlake.yml [hunk 6]
```diff
@@ -67,24 +94,15 @@
...
+          url: /learn/testing-ballerina-code/executing-tests/#test-execution-behavior
+          active: test-execution-behavior
+        - title: behavior During Failures
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:103]:
```suggestion
        - title: Behavior During Failures
```

File: _data/guidesnavswanlake.yml [hunk 7]
```diff
@@ -39,6 +39,21 @@
...
```
Comment (praneesha) [inline _data/guidesnavswanlake.yml:42]:
```suggestion
    - title: Structuring the Tests
```
Please check and make sure all the titles are in title case.

File: _data/guidesnavswanlake.yml [hunk 8]
```diff
@@ -39,6 +39,21 @@
...
+      url: /learn/testing-ballerina-code/structuring-the-tests/
+      active: structuring-the-tests
+    - title: Configuring Tests
```
Comment (praneesha) [inline _data/guidesnavswanlake.yml:45]:
```suggestion
    - title: Configuring Tests
```
If we say `Structuring the Tests`, this should also be `Configuring the Tests`. Please make the titles consistent.

File: _data/guidesnavswanlake.yml [hunk 9]
```diff
@@ -39,6 +39,21 @@
...
+      url: /learn/testing-ballerina-code/configuring-tests/
+      active: configuring-tests
+    - title: Code coverage and Reporting
```
Comment (praneesha) [inline _data/guidesnavswanlake.yml:48]:
```suggestion
    - title: Code Coverage and Reporting
```

File: _data/guidesnavswanlake.yml [hunk 10]
```diff
@@ -39,6 +39,21 @@
...
+      url: /learn/testing-ballerina-code/code-coverage-and-reporting/
+      active: code-coverage-and-reporting
+    - title: Data-driven Tests
+      url: /learn/testing-ballerina-code/data-driven-tests/
+      active: data-drive-tests
+    - title: Test groups
```
Comment (praneesha) [inline _data/guidesnavswanlake.yml:54]:
```suggestion
    - title: Test Groups
```

File: _data/testingcodeswanlake.yml [hunk 11]
```diff
@@ -7,45 +7,72 @@
...
-        - title: Project Structure
-          url: /learn/testing-ballerina-code/writing-tests/#project-structure
-          active: project-structure
-        - title: Test Source Files
-          url: /learn/testing-ballerina-code/writing-tests/#test-source-files
-          active: test-source-files
-        - title: Test Resources
-          url: /learn/testing-ballerina-code/writing-tests/#test-resources
-          active: test-resources
         - title: Defining a Test
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:13]:
```suggestion
        - title: Defining a Test
```
This should also be `Defining Tests` to make the titles consistent.

File: _data/testingcodeswanlake.yml [hunk 12]
```diff
@@ -7,45 +7,72 @@
...
         - title: Using Assertions
           url: /learn/testing-ballerina-code/writing-tests/#using-assertions
           active: using-assertions
-        - title: Setup and Teardown
-          url: /learn/testing-ballerina-code/writing-tests/#setup-and-teardown
-          active: setup-and-teardown
-        - title: test:AfterGroups {}
-          url: /learn/testing-ballerina-code/writing-tests/#testaftergroups-
-          active: testaftergroups-
+    - title: Structuring the tests
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:19]:
```suggestion
    - title: Structuring the Tests
```

File: _data/testingcodeswanlake.yml [hunk 13]
```diff
@@ -7,45 +7,72 @@
...
-        - title: What’s Next?
-          url: /learn/testing-ballerina-code/writing-tests/#whats-next
-          active: whats-next
+    - title: Configuring Tests
+      url : '#'
+      sublinks:
+        - title: Setup and Teardown
+          url: /learn/testing-ballerina-code/configuring-tests/#setup-and-teardown
+          active: setup-and-teardown
+        - title: Test specific Configurations
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:37]:
```suggestion
        - title: Test Specific Configurations
```

File: _data/testingcodeswanlake.yml [hunk 14]
```diff
@@ -7,45 +7,72 @@
...
+        - title: Test Report
+          url: /learn/testing-ballerina-code/code-coverage-and-reporting/#test-report
+          active: test-report
+        - title: Code Coverage Report
+          url: /learn/testing-ballerina-code/code-coverage-and-reporting/#code-coverage-report
+          active: code-coverage-report
+        - title: JaCoCo XML Report
+          url: /learn/testing-ballerina-code/code-coverage-and-reporting/#jacoco-xml-report
+          active: jacoco-xml-report
+    - title: Data-driven Tests
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:58]:
```suggestion
    - title: Data-Driven Tests
```

File: _data/testingcodeswanlake.yml [hunk 15]
```diff
@@ -7,45 +7,72 @@
...
+      url : '#'
+      sublinks:
+        - title: Using data providers
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:61]:
```suggestion
        - title: Using Data Providers
```

File: _data/testingcodeswanlake.yml [hunk 16]
```diff
@@ -7,45 +7,72 @@
...
+          url: /learn/testing-ballerina-code/data-driven-tests/#using-data-providers
+          active: using-data-providers
+        - title: Executing specific data sets
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:64]:
```suggestion
        - title: Executing Specific Data Sets
```

File: _data/testingcodeswanlake.yml [hunk 17]
```diff
@@ -7,45 +7,72 @@
...
+          url: /learn/testing-ballerina-code/data-driven-tests/#executing-specific-data-sets
+          active: executing-specific-data-sets
+    - title: Test groups
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:67]:
```suggestion
    - title: Test Groups
```

File: _data/testingcodeswanlake.yml [hunk 18]
```diff
@@ -7,45 +7,72 @@
...
+      url: "#"
+      sublinks:
+        - title: Grouping tests
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:70]:
```suggestion
        - title: Grouping Tests
```

File: _data/testingcodeswanlake.yml [hunk 19]
```diff
@@ -7,45 +7,72 @@
...
+          url: /learn/testing-ballerina-code/test-groups/#grouping-tests
+          active: grouping-tests
+        - title: Grouping commands
```
Comment (praneesha) [inline _data/testingcodeswanlake.yml:73]:
```suggestion
        - title: Grouping Commands
```

File: learn/guides/testing-ballerina-code/code-coverage-and-reporting.md [hunk 20]
```diff
@@ -0,0 +1,68 @@
+---
+layout: ballerina-testing-code-left-nav-pages-swanlake
+title: Code Coverage and Reporting
+description: Learn how to do code coverage and reporting for ballerina tests.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/code-coverage-and-reporting.md:4]:
```suggestion
description: Learn how to perform code coverage and reporting for Ballerina tests.
```

File: learn/guides/testing-ballerina-code/code-coverage-and-reporting.md [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+keywords: ballerina, programming language, testing
+permalink: /learn/testing-ballerina-code/code-coverage-and-reporting/
+active: code-coverage-and-reporting
+intro: The Ballerina Test Framework follows a general organized structure that allows for testing code under various conditions by making use of resources and configurations.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/code-coverage-and-reporting.md:8]:
```suggestion
intro: The Ballerina Test Framework follows a general, organized structure that allows for testing code under various conditions by making use of resources and configurations. 
```

File: learn/guides/testing-ballerina-code/code-coverage-and-reporting.md [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+$ bal test --test-report --code-coverage
+```
+
+A sample view of the code coverage report is shown below.
+
+![Sample Code Coverage](/learn/images/code-cov.gif)
+
+## JaCoCo XML Report
+
+You can pass the `--coverage-format=xml`  flag along with the `--code-coverage` flag in the test execution command and
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/code-coverage-and-reporting.md:65]:
```suggestion
You can pass the `--coverage-format=xml` flag along with the `--code-coverage` flag in the test execution command and
```

File: learn/guides/testing-ballerina-code/code-coverage-and-reporting.md [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+ generate the JaCoCo XML report at the end of the test execution.
+ This file can be uploaded to CI/CD tools (e.g., CodeCov) to display the coverage information for both Ballerina and
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/code-coverage-and-reporting.md:67]:
```suggestion
 This file can be uploaded to the CI/CD tools (e.g., CodeCov) to display the coverage information for both Ballerina and
```

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 24]
```diff
@@ -0,0 +1,267 @@
...
+- /learn/user-guide/testing-ballerina-code/configuring-tests/
+---
+
+## Setup and Teardown
+The following test annotations can be used to set up and tear down the instructions. These configuration annotations 
+enable executing instructions at various levels.
+
+### Suite Level
+
+#### @test:BeforeSuite {}
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:27]:
```suggestion
#### `@test:BeforeSuite {}`
```
Please check and add backticks in all the occurrences.

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 25]
```diff
@@ -0,0 +1,267 @@
...
+The `BeforeEach` annotated function will be run before each test in the test suite. This can be used to initialize the 
+test-level prerequisites repeatedly before every test function.
+
+***Example:***
+
+```ballerina
+import ballerina/io;
+import ballerina/test;
+
+// The `BeforeEach` function, which is executed before each test function
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:172]:
```suggestion
// The `BeforeEach` function, which is executed before each test function.
```
Please check the full stop in all occurrences.

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 26]
```diff
@@ -0,0 +1,267 @@
...
+    io:println("I'm in test function 3!");
+    test:assertTrue(true, msg = "Failed!");
+}
+```
+
+### Each Test Case
+
+#### @test:Config { before : “ “ }
+The test config annotation makes use of ‘before’ to denote which function needs to execute before the particular 
+test is run
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:242]:
```suggestion
test is run.
```

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 27]
```diff
@@ -0,0 +1,267 @@
...
+
+#### @test:Config { after : “ “ }
+The test config annotation makes use of ‘after’ to denote which function needs to execute before the particular 
+test is run
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:246]:
```suggestion
test is run.
```
Please check the full stop in all the occurrences.

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 28]
```diff
@@ -0,0 +1,267 @@
...
+
+
+## Test Specific Configurations
+Configurations for testing can be provided using configurable variables. The values for configurable variables can be
+provided in a file named `Config.toml` located in the `tests/` directory which will only be initialized when the tests
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:251]:
```suggestion
provided in a file named `Config.toml` located in the `tests/` directory, which will only be initialized when the tests
```

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 29]
```diff
@@ -0,0 +1,267 @@
...
+are run. It is important to note that if the `Config.toml` is not specified in the `tests/` directory, the values will
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:252]:
```suggestion
are run. 

>**Note:** If the `Config.toml` is not specified in the `tests/` directory, the values will
```
Shall we make this a separate `Note` section?

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 30]
```diff
@@ -0,0 +1,267 @@
...
+be either taken from the `Config.toml` in the root directory or default values will be used.
+
+Configurable variables are useful when we require separate configurations that cannot be feasibly used outside of
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:255]:
```suggestion
Configurable variables are useful when you require separate configurations that cannot be feasibly used outside of 
```

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 31]
```diff
@@ -0,0 +1,267 @@
...
+
+
+## Test Specific Configurations
+Configurations for testing can be provided using configurable variables. The values for configurable variables can be
+provided in a file named `Config.toml` located in the `tests/` directory which will only be initialized when the tests
+are run. It is important to note that if the `Config.toml` is not specified in the `tests/` directory, the values will
+be either taken from the `Config.toml` in the root directory or default values will be used.
+
+Configurable variables are useful when we require separate configurations that cannot be feasibly used outside of 
+testing. This is particularly useful when testing services and clients where we may need different host values when we
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:256]:
```suggestion
testing. This is particularly useful when testing services and clients where you may need different host values when you
```
Please check all occurences.

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 32]
```diff
@@ -0,0 +1,267 @@
...
+are trying to test the service or client.
+
+
+## Test-only Dependencies
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:260]:
```suggestion
## Test-Only Dependencies
```
Please check all titles.

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 33]
```diff
@@ -0,0 +1,267 @@
...
+Dependencies meant to be resolved only during testing and can be specified in the Ballerina.toml file by specifying the scope
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:261]:
```suggestion
Dependencies meant to be resolved only during testing and can be specified in the `Ballerina.toml` file by specifying the scope.
```

File: learn/guides/testing-ballerina-code/configuring-tests.md [hunk 34]
```diff
@@ -0,0 +1,267 @@
...
+
+    ```
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/configuring-tests.md:263]:
```suggestion
    ```toml
```
Please check all code snippets.

File: learn/guides/testing-ballerina-code/data-driven-tests.md [hunk 35]
```diff
@@ -0,0 +1,157 @@
...
+- /learn/user-guide/testing-ballerina-code/data-driven-tests
+- /learn/user-guide/testing-ballerina-code/data-driven-tests/
+- /learn/user-guide/testing-ballerina-code/
+- /learn/user-guide/testing-ballerina-code
+- /learn/user-guide/testing-ballerina-code/data-driven-tests/
+---
+
+## Using Data Providers
+
+A data provider is a function that will be used to provide the data sets for a test function.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/data-driven-tests.md:23]:
```suggestion
A data provider is a function, which will be used to provide the data sets for a test function.
```

File: learn/guides/testing-ballerina-code/data-driven-tests.md [hunk 36]
```diff
@@ -0,0 +1,157 @@
...
+
+                3 passing
+                0 failing
+                0 skipped
+```
+
+## Executing Ppecific Data Sets
+
+If you need to run only a specific case from the given data set, you can use the test name with the key to do that.
+You can make use of wild cards(`*`) to capture multiple cases as well.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/data-driven-tests.md:117]:
```suggestion
You can make use of wild cards(`*`) to capture multiple cases as well.
```
Shall we make this an `Info` to a `Tip` section?

File: learn/guides/testing-ballerina-code/data-driven-tests.md [hunk 37]
```diff
@@ -0,0 +1,157 @@
...
+		[pass] fruitsDataProviderTest#'banana'
+
+		1 passing
+		0 failing
+		0 skipped
+```
+
+>**Note:** Make sure to include the key within double quotes. 
+
+***Example for array of array data sets:***
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/data-driven-tests.md:140]:
```suggestion
***Example for an array of array data sets:***
```
Make this consistent with the other occurrences.

File: learn/guides/testing-ballerina-code/executing-tests.md [hunk 38]
```diff
@@ -14,37 +14,72 @@ redirect_from:
...
+
+![Test Execution Order](/learn/images/test-execution-order.png)
+
+
+## Behavior During Failures
+
+* If BeforeSuite function fails,
+ every other following function is skipped.
+
+* If a BeforeGroups function fails,
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/executing-tests.md:29]:
```suggestion
* If a `BeforeGroups` function fails,
```

File: learn/guides/testing-ballerina-code/executing-tests.md [hunk 39]
```diff
@@ -14,37 +14,72 @@ redirect_from:
...
+ the tests belonging to that group, setup and teardown functions specific to those tests will fail.
+ The AfterGroups functions for that test group will be skipped. Other setup and teardown functions will be executed as expected.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/executing-tests.md:31]:
```suggestion
 The `AfterGroups` functions for that test group will be skipped. Other setup and teardown functions will be executed as expected.
```

File: learn/guides/testing-ballerina-code/executing-tests.md [hunk 40]
```diff
@@ -14,37 +14,72 @@ redirect_from:
...
+ Tests belonging to other groups will not be affected.
+
+* If BeforeEach function fails,
+Every test will be skipped. Since BeforeSuite is already executed, AfterSuite will be executed.
+ Even though the BeforeGroups function is executed prior to BeforeEach, the AfterGroups function will not be
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/executing-tests.md:36]:
```suggestion
 Even though the `eforeGroups` function is executed prior to `BeforeEach`, the `AfterGroups` function will not be
```

File: learn/guides/testing-ballerina-code/executing-tests.md [hunk 41]
```diff
@@ -14,37 +14,72 @@ redirect_from:
...
+ executed unless marked as `alwaysRun`.
+
+* If a test function fails, none of the other functions get skipped.
+
+* If the before test function fails, the test function and the after function for that test will be skipped.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/executing-tests.md:41]:
```suggestion
* If the before test function fails, the test function and the after function for that test will be skipped.
```
Should this be `BeforeTest` to be consistent?

File: learn/guides/testing-ballerina-code/executing-tests.md [hunk 42]
```diff
@@ -14,37 +14,72 @@ redirect_from:
...
+
+* If AfterEach function fails, every following BeforeEach, AfterEach and test function will get skipped.
+
+* If `alwaysRun` property is enabled, AfterGroups & AfterSuite functions will run  irrespective of the status of other functions.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/executing-tests.md:45]:
```suggestion
* If an `alwaysRun` property is enabled, `the AfterGroups` and `AfterSuite` functions will run irrespective of the status of the other functions.
```

File: learn/guides/testing-ballerina-code/structuring-the-tests.md [hunk 43]
```diff
@@ -0,0 +1,42 @@
+---
+layout: ballerina-testing-code-left-nav-pages-swanlake
+title: Structuring the Tests
+description: Learn how to structure Ballerina tests.
+keywords: ballerina, programming language, testing
+permalink: /learn/testing-ballerina-code/structuring-the-tests/
+active: structuring-the-tests
+intro: The Ballerina Test Framework follows a general organized structure that allows for testing code under various conditions by making use of resources and configurations.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/structuring-the-tests.md:8]:
```suggestion
intro: The Ballerina Test Framework follows a general, organized structure that allows testing code under various conditions by making use of resources and configurations. 
```
Please check and correct all occurrencs.

File: learn/guides/testing-ballerina-code/structuring-the-tests.md [hunk 44]
```diff
@@ -0,0 +1,42 @@
...
+- /learn/user-guide/testing-ballerina-code/structuring-the-tests
+- /learn/user-guide/testing-ballerina-code/structuring-the-tests/
+- /learn/user-guide/testing-ballerina-code/
+- /learn/user-guide/testing-ballerina-code
+- /learn/user-guide/testing-ballerina-code/structuring-the-tests/
+---
+
+## Test Source Files
+Unit tests bound to a module need to be placed in a sub folder called `tests/` within the module. 
+In a standard Ballerina package, a module is mapped to a test suite. All tests within a module’s `tests/` subfolder are
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/structuring-the-tests.md:23]:
```suggestion
In a standard Ballerina package, a module is mapped to a test suite. All tests within a module’s `tests/` subdirectory are 
```
Use either `folder` or `directory` consistently.

File: learn/guides/testing-ballerina-code/structuring-the-tests.md [hunk 45]
```diff
@@ -0,0 +1,42 @@
...
+considered to be part of the same test suite. The test source files could have any name. The test functions are just 
+Ballerina functions, which use a special annotation to mark the function as a test. Test functions must be specified 
+with the `@test:Config {}` annotation and there is no restriction on the test function name.
+
+The functions, services, and the global variables defined in a module are accessible from within the test files.
+Hence, you cannot redefine a symbol in the test files if it is already declared in the module.
+On the other hand, symbols defined in the test files will not be visible inside the module source files.
+
+## Test Resources
+The `resources/` subdirectory found within the `tests/` directory is meant to contain any files or resources that are
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/structuring-the-tests.md:33]:
```suggestion
The `resources/` sub directory found within the `tests/` directory is meant to contain any files or resources that are 
```

File: learn/guides/testing-ballerina-code/test-groups.md [hunk 46]
```diff
@@ -0,0 +1,123 @@
...
+- /learn/user-guide/testing-ballerina-code/
+- /learn/user-guide/testing-ballerina-code
+- /learn/user-guide/testing-ballerina-code/test-groups/
+---
+
+## Grouping Tests
+Test grouping allows us to control the execution of tests by partitioning them into groups. The Test Framework allows
+for a single test to have multiple groups. 
+
+A test is assigned to a group via the test configuration `groups` key which accepts a string array of groups.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/test-groups.md:25]:
```suggestion
A test is assigned to a group via the test configuration `groups` key, which accepts a string array of groups. 
```

File: learn/guides/testing-ballerina-code/test-groups.md [hunk 47]
```diff
@@ -0,0 +1,123 @@
...
+
+                [pass] testFunction1
+                [pass] testFunction2
+
+                2 passing
+                0 failing
+                0 skipped
+```
+
+The groups flag supports executing multiple groups by using comma separated arguments
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/test-groups.md:77]:
```suggestion
The `groups` flag supports executing multiple groups by using comma-separated arguments.
```

File: learn/guides/testing-ballerina-code/test-groups.md [hunk 48]
```diff
@@ -0,0 +1,123 @@
...
+                [pass] testFunction3
+
+                3 passing
+                0 failing
+                0 skipped
+```
+
+### Disabling Grouped Tests
+
+To skip execution of certain groups of tests, the `--disable-groups` flag is used during test execution.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/test-groups.md:103]:
```suggestion
To skip the execution of certain groups of tests, the `--disable-groups` flag is used during test execution.
```

File: learn/guides/testing-ballerina-code/test-groups.md [hunk 49]
```diff
@@ -0,0 +1,123 @@
...
+
+```bal test --disable-groups g2```
+
+which results in an output of :
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/test-groups.md:107]:
```suggestion
The above results in an output of :
```

File: learn/guides/testing-ballerina-code/writing-tests.md [hunk 50]
```diff
@@ -57,15 +29,23 @@ Once the test module is imported, the following annotation can be used to write
 
 ### '@test:Config {}'
 
-The function specified after the annotation is a test function. This annotation supports the following value fields.
+The function specified after the annotation is a test function. This annotation supports fields to configure the test case with the following.
 
+*   Enable/disable the test.
+    ***enable: {true&#124;false}*** - Default: true
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/writing-tests.md:35]:
```suggestion
    ***`enable: {true&#124;false}`*** - The default value is `true`.
```
Please check backticks in all occurrences.

File: learn/guides/testing-ballerina-code/writing-tests.md [hunk 51]
```diff
@@ -57,15 +29,23 @@ Once the test module is imported, the following annotation can be used to write
...
+    
+*   Set up and tear down prerequisites required for a test case.
+    ***before: &lt;function name&gt;*** - The function to be run just before the test is run. Default: none
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/writing-tests.md:38]:
```suggestion
    ***before: &lt;function name&gt;*** - The function to be run just before the test is run. The default value is `none`.
```

File: learn/guides/testing-ballerina-code/writing-tests.md [hunk 52]
```diff
@@ -57,15 +29,23 @@ Once the test module is imported, the following annotation can be used to write
...
+    ***after: &lt;function name&gt;*** - The function to be run just after the test is run. Default: none
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/writing-tests.md:39]:
```suggestion
    ***after: &lt;function name&gt;*** - The function to be run just after the test is run. The default value is `none`.
```

File: learn/guides/testing-ballerina-code/writing-tests.md [hunk 53]
```diff
@@ -57,15 +29,23 @@ Once the test module is imported, the following annotation can be used to write
...
+    
+*   Define the order of test execution.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/writing-tests.md:41]:
```suggestion
*   Defines the order of the test execution.
```

File: learn/guides/testing-ballerina-code/writing-tests.md [hunk 54]
```diff
@@ -57,15 +29,23 @@ Once the test module is imported, the following annotation can be used to write
...
 
-*   ***enable: {true&#124;false}*** - Enable/disable the test. Default: true
-*   ***before: &lt;function name&gt;*** - The function to be run just before the test is run. Default: none
-*   ***after: &lt;function name&gt;*** - The function to be run just after the test is run. Default: none
-*   ***dependsOn: [&lt;function names>, …]*** - List of functions on which the test function depends. The order in which the comma-separated list appears has no prominence. In case there needs to be an order, the `dependsOn` parameter can be used to create an ordered sequence of functions with one function depending on the other.
-*   ***dataProvider: &lt;function name>*** - Specifies the function that will be used to provide the data sets for the test.
-*   ***groups: [“&lt;test group name”, …]*** - A comma-separated list of test group names (one or more) to which this test
+*   Define data sets for data-driven tests.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/writing-tests.md:44]:
```suggestion
*   Define data sets for data-driven tests.
```
Are these titles? Better if we can make them sub titles.

File: learn/guides/testing-ballerina-code/writing-tests.md [hunk 55]
```diff
@@ -57,15 +29,23 @@ Once the test module is imported, the following annotation can be used to write
...
+    ***dataProvider: &lt;function name>*** - Specifies the function that will be used to provide the data sets for the test.
```
Comment (praneesha) [inline learn/guides/testing-ballerina-code/writing-tests.md:45]:
```suggestion
    ***dataProvider: &lt;function name>*** - Specifies the function, which will be used to provide the data sets for the test.
```

File: _data/guidesnavswanlake.yml (hunk 8 above)
Comment (Ibaqu) [reply _data/guidesnavswanlake.yml:45]: Changed `Structuring the Tests` to `Structuring Tests`

File: learn/guides/testing-ballerina-code/configuring-tests.md (hunk 31 above)
Comment (Ibaqu) [reply learn/guides/testing-ballerina-code/configuring-tests.md]: Changed `we` to `you`

File: learn/guides/testing-ballerina-code/data-driven-tests.md (hunk 36 above)
Comment (Ibaqu) [reply learn/guides/testing-ballerina-code/data-driven-tests.md:117]: I changed this into a tip section but it didnt look as good when generating the website. Would it be possible to just include it in this paragraph itself without another tip section? @praneesha

Comment (praneesha) [reply learn/guides/testing-ballerina-code/data-driven-tests.md:117]: +1

**Review by praneesha — APPROVED** [review]

**Pull request merged**